File: server/variable_resolver.py
"""
Variable Resolution Utility

Provides generic functionality to resolve variables embedded in text content.
Variables are marked with #{VariableName} format and are replaced with actual values
from the provided user_data dictionary.

This utility is designed to be flexible and reusable across different contexts
throughout the application (messages, templates, notifications, etc.).
"""

import re
from typing import Dict, Any, List, Optional, Callable, Union


class VariableResolverOptions:
    """Configuration options for variable resolution behavior."""
    
    def __init__(
        self,
        default_value: Optional[str] = None,
        remove_unresolved: bool = False,
        preserve_format: bool = False,
        formatter: Optional[Callable[[str, Any], str]] = None,
        case_sensitive: bool = False
    ):
        """
        Initialize variable resolver options.
        
        Args:
            default_value: The default value to use when a variable is not found.
                          If not provided, unresolved variables will be left as-is.
            remove_unresolved: Whether to remove the variable placeholder if not found.
                              If True, #{UnknownVar} will be removed from the text.
            preserve_format: Whether to preserve the variable format (#{}) when using defaultValue.
            formatter: Custom formatter function to transform variable values before insertion.
            case_sensitive: Whether variable matching should be case-sensitive.
                           Default: False (case-insensitive)
        """
        self.default_value = default_value
        self.remove_unresolved = remove_unresolved
        self.preserve_format = preserve_format
        self.formatter = formatter
        self.case_sensitive = case_sensitive


# Regular expression to match variable placeholders in text.
# Matches patterns like #{VariableName}, #{Variable_Name}, #{Variable123}
#
# Pattern breakdown:
# - #\{ - matches literal "#{"
# - ([a-zA-Z_][a-zA-Z0-9_]*) - captures variable name (must start with letter or underscore)
# - \} - matches literal "}"
VARIABLE_PATTERN = re.compile(r'#\{([a-zA-Z_][a-zA-Z0-9_]*)\}')


def resolve_variables(
    text: str,
    user_data: Dict[str, Any],
    options: Optional[Union[VariableResolverOptions, Dict[str, Any]]] = None
) -> str:
    """
    Resolve all variables in the given text with actual values from user_data.
    
    Args:
        text: The text containing variable placeholders (e.g., "Hello #{Name}!")
        user_data: Dictionary containing variable values (e.g., {"Name": "John"})
        options: Optional configuration for resolution behavior (VariableResolverOptions or dict)
    
    Returns:
        The text with all variables resolved to their actual values
    
    Examples:
        >>> resolve_variables("Hello #{Name}!", {"Name": "Alice"})
        'Hello Alice!'
        
        >>> resolve_variables(
        ...     "Hi #{Name}, your email is #{Email}",
        ...     {"Name": "Bob"},
        ...     VariableResolverOptions(default_value="[Not provided]")
        ... )
        'Hi Bob, your email is [Not provided]'
        
        >>> resolve_variables(
        ...     "Hello #{Name} #{UnknownVar}!",
        ...     {"Name": "Charlie"},
        ...     VariableResolverOptions(remove_unresolved=True)
        ... )
        'Hello Charlie !'
    """
    if not text:
        return text
    
    # Convert dict to VariableResolverOptions if needed
    if options is None:
        options = VariableResolverOptions()
    elif isinstance(options, dict):
        options = VariableResolverOptions(**options)
    
    # Create a case-insensitive lookup if needed
    if not options.case_sensitive:
        data_lookup = {k.lower(): k for k in user_data.keys()}
    else:
        data_lookup = {k: k for k in user_data.keys()}
    
    def replace_variable(match: re.Match) -> str:
        """Replace a single variable match with its value."""
        full_match = match.group(0)
        variable_name = match.group(1)
        
        # Get the actual key (with original casing) from user data
        lookup_key = variable_name if options.case_sensitive else variable_name.lower()
        actual_key = data_lookup.get(lookup_key)
        
        # Check if variable exists in user data
        if actual_key and actual_key in user_data:
            value = user_data[actual_key]
            
            # Handle None values
            if value is None:
                if options.remove_unresolved:
                    return ''
                if options.default_value is not None:
                    return full_match if options.preserve_format else options.default_value
                return full_match  # Keep original placeholder
            
            # Apply custom formatter if provided
            if options.formatter:
                return options.formatter(actual_key, value)
            
            # Convert value to string
            return str(value)
        
        # Variable not found - handle based on options
        if options.remove_unresolved:
            return ''
        
        if options.default_value is not None:
            return full_match if options.preserve_format else options.default_value
        
        # Keep original placeholder
        return full_match
    
    return VARIABLE_PATTERN.sub(replace_variable, text)


def extract_variable_names(text: str) -> List[str]:
    """
    Extract all variable names from text (without resolving them).
    Useful for determining which variables are used in a template.
    
    Args:
        text: The text to extract variables from
    
    Returns:
        List of unique variable names found in the text
    
    Examples:
        >>> extract_variable_names("Hello #{Name}, your email is #{Email} and name is #{Name}")
        ['Name', 'Email']
    """
    if not text:
        return []
    
    matches = VARIABLE_PATTERN.findall(text)
    # Return unique variable names (preserve order)
    seen = set()
    unique_vars = []
    for var in matches:
        if var not in seen:
            seen.add(var)
            unique_vars.append(var)
    return unique_vars


def has_variables(text: str) -> bool:
    """
    Check if the given text contains any variable placeholders.
    
    Args:
        text: The text to check
    
    Returns:
        True if text contains at least one variable placeholder
    
    Examples:
        >>> has_variables("Hello #{Name}!")
        True
        >>> has_variables("Hello World!")
        False
    """
    if not text:
        return False
    
    return VARIABLE_PATTERN.search(text) is not None


def validate_variables(text: str) -> Dict[str, Union[bool, List[str]]]:
    """
    Validate variable names in text to ensure they follow naming conventions.
    Variable names must start with a letter or underscore and contain only
    letters, numbers, and underscores.
    
    Args:
        text: The text to validate
    
    Returns:
        Dictionary with validation result and any invalid variable names found
    
    Examples:
        >>> validate_variables("Hello #{Name} and #{123Invalid}")
        {'valid': False, 'invalid_vars': ['123Invalid']}
    """
    # This pattern matches ANY #{...} including invalid ones
    any_variable_pattern = re.compile(r'#\{([^}]+)\}')
    valid_name_pattern = re.compile(r'^[a-zA-Z_][a-zA-Z0-9_]*$')
    
    invalid_vars = []
    matches = any_variable_pattern.findall(text)
    
    for var_name in matches:
        if not valid_name_pattern.match(var_name):
            invalid_vars.append(var_name)
    
    return {
        'valid': len(invalid_vars) == 0,
        'invalid_vars': invalid_vars
    }


def preview_text(text: str, sample_data: Optional[Dict[str, Any]] = None) -> str:
    """
    Preview text with variables resolved, useful for showing users
    what the final text will look like with sample data.
    
    Args:
        text: The text template
        sample_data: Sample variable values for preview
    
    Returns:
        Resolved text with indication of missing variables
    
    Examples:
        >>> preview_text("Welcome #{Name}, email: #{Email}", {"Name": "Alice"})
        'Welcome Alice, email: [Not set]'
    """
    if sample_data is None:
        sample_data = {}
    
    return resolve_variables(
        text,
        sample_data,
        VariableResolverOptions(
            default_value='[Not set]',
            remove_unresolved=False
        )
    )


def batch_resolve_variables(
    texts: List[str],
    user_data: Dict[str, Any],
    options: Optional[Union[VariableResolverOptions, Dict[str, Any]]] = None
) -> List[str]:
    """
    Batch resolve variables in multiple text strings.
    Useful for resolving variables in a set of messages or templates at once.
    
    Args:
        texts: List of text strings to resolve
        user_data: Variable values to use for resolution
        options: Resolution options
    
    Returns:
        List of resolved text strings
    
    Examples:
        >>> batch_resolve_variables(
        ...     ["Hello #{Name}", "Email: #{Email}"],
        ...     {"Name": "Bob", "Email": "bob@example.com"}
        ... )
        ['Hello Bob', 'Email: bob@example.com']
    """
    return [resolve_variables(text, user_data, options) for text in texts]


# Convenience function for common use case: simple resolution with defaults
def simple_resolve(text: str, user_data: Dict[str, Any], default: str = '') -> str:
    """
    Simple variable resolution with a default value for missing variables.
    
    Args:
        text: Text containing variables
        user_data: Variable values
        default: Default value for missing variables (default: empty string)
    
    Returns:
        Resolved text
    
    Examples:
        >>> simple_resolve("Hello #{Name}!", {"Name": "World"})
        'Hello World!'
        >>> simple_resolve("Hello #{Name}!", {}, "[Guest]")
        'Hello [Guest]!'
    """
    return resolve_variables(
        text,
        user_data,
        VariableResolverOptions(default_value=default)
    )

File: client/src/utils/variableResolver.ts
/**
 * Variable Resolution Utility
 * 
 * Provides generic functionality to resolve variables embedded in text content.
 * Variables are marked with #{VariableName} format and are replaced with actual values
 * from the provided user_data object.
 * 
 * This utility is designed to be flexible and reusable across different contexts
 * throughout the application (messages, templates, notifications, etc.).
 */

export interface VariableResolverOptions {
    /**
     * The default value to use when a variable is not found in user_data.
     * If not provided, unresolved variables will be left as-is in the text.
     */
    defaultValue?: string;
    
    /**
     * Whether to remove the variable placeholder if it's not found.
     * If true, #{UnknownVar} will be removed from the text.
     * If false and no defaultValue is provided, #{UnknownVar} stays as-is.
     */
    removeUnresolved?: boolean;
    
    /**
     * Whether to preserve the variable format (#{}) when using defaultValue.
     * If true, unresolved variables will be shown as "#{VarName}" with defaultValue.
     * If false, only the defaultValue will be shown.
     */
    preserveFormat?: boolean;
    
    /**
     * Custom formatter function to transform variable values before insertion.
     * Useful for formatting dates, numbers, or applying custom transformations.
     */
    formatter?: (key: string, value: any) => string;
    
    /**
     * Case sensitivity for variable matching.
     * If true, #{Name} and #{name} are treated as different variables.
     * Default: false (case-insensitive)
     */
    caseSensitive?: boolean;
}

/**
 * Regular expression to match variable placeholders in text.
 * Matches patterns like #{VariableName}, #{Variable_Name}, #{Variable123}
 * 
 * Pattern breakdown:
 * - #\{ - matches literal "#{"
 * - ([a-zA-Z_][a-zA-Z0-9_]*) - captures variable name (must start with letter or underscore)
 * - \} - matches literal "}"
 * - g flag - global matching to find all occurrences
 */
const VARIABLE_PATTERN = /#\{([a-zA-Z_][a-zA-Z0-9_]*)\}/g;

/**
 * Resolves all variables in the given text with actual values from user_data.
 * 
 * @param text - The text containing variable placeholders (e.g., "Hello #{Name}!")
 * @param userData - Object containing variable values (e.g., { Name: "John" })
 * @param options - Optional configuration for resolution behavior
 * @returns The text with all variables resolved to their actual values
 * 
 * @example
 * // Basic usage
 * resolveVariables("Hello #{Name}!", { Name: "Alice" })
 * // Returns: "Hello Alice!"
 * 
 * @example
 * // With default value for missing variables
 * resolveVariables("Hi #{Name}, your email is #{Email}", 
 *   { Name: "Bob" }, 
 *   { defaultValue: "[Not provided]" }
 * )
 * // Returns: "Hi Bob, your email is [Not provided]"
 * 
 * @example
 * // Remove unresolved variables
 * resolveVariables("Hello #{Name} #{UnknownVar}!", 
 *   { Name: "Charlie" }, 
 *   { removeUnresolved: true }
 * )
 * // Returns: "Hello Charlie !"
 * 
 * @example
 * // With custom formatter
 * resolveVariables("Order total: #{Total}", 
 *   { Total: 1234.56 },
 *   { formatter: (key, value) => key === 'Total' ? `$${value.toFixed(2)}` : value }
 * )
 * // Returns: "Order total: $1234.56"
 */
export function resolveVariables(
    text: string,
    userData: Record<string, any>,
    options: VariableResolverOptions = {}
): string {
    if (!text) {
        return text;
    }
    
    const {
        defaultValue,
        removeUnresolved = false,
        preserveFormat = false,
        formatter,
        caseSensitive = false
    } = options;
    
    // Create a case-insensitive lookup map if needed
    const dataLookup = caseSensitive
        ? userData
        : createCaseInsensitiveLookup(userData);
    
    return text.replace(VARIABLE_PATTERN, (match, variableName) => {
        // Get the actual key (with original casing) from user data
        const lookupKey = caseSensitive 
            ? variableName 
            : variableName.toLowerCase();
        
        const actualKey = caseSensitive
            ? variableName
            : Object.keys(userData).find(k => k.toLowerCase() === lookupKey);
        
        // Check if variable exists in user data
        if (actualKey && actualKey in userData) {
            const value = userData[actualKey];
            
            // Handle null or undefined values
            if (value === null || value === undefined) {
                if (removeUnresolved) {
                    return '';
                }
                if (defaultValue !== undefined) {
                    return preserveFormat ? match : defaultValue;
                }
                return match; // Keep original placeholder
            }
            
            // Apply custom formatter if provided
            if (formatter) {
                return formatter(actualKey, value);
            }
            
            // Convert value to string
            return String(value);
        }
        
        // Variable not found - handle based on options
        if (removeUnresolved) {
            return '';
        }
        
        if (defaultValue !== undefined) {
            return preserveFormat ? match : defaultValue;
        }
        
        // Keep original placeholder
        return match;
    });
}

/**
 * Creates a case-insensitive lookup map from user data.
 * This allows matching #{Name}, #{name}, #{NAME} to the same variable.
 */
function createCaseInsensitiveLookup(userData: Record<string, any>): Record<string, any> {
    const lookup: Record<string, any> = {};
    for (const key in userData) {
        lookup[key.toLowerCase()] = userData[key];
    }
    return lookup;
}

/**
 * Extracts all variable names from text (without resolving them).
 * Useful for determining which variables are used in a template.
 * 
 * @param text - The text to extract variables from
 * @returns Array of unique variable names found in the text
 * 
 * @example
 * extractVariableNames("Hello #{Name}, your email is #{Email} and name is #{Name}")
 * // Returns: ["Name", "Email"]
 */
export function extractVariableNames(text: string): string[] {
    if (!text) {
        return [];
    }
    
    const variables = new Set<string>();
    const matches = Array.from(text.matchAll(VARIABLE_PATTERN));
    
    for (const match of matches) {
        variables.add(match[1]);
    }
    
    return Array.from(variables);
}

/**
 * Checks if the given text contains any variable placeholders.
 * 
 * @param text - The text to check
 * @returns true if text contains at least one variable placeholder
 * 
 * @example
 * hasVariables("Hello #{Name}!")  // Returns: true
 * hasVariables("Hello World!")    // Returns: false
 */
export function hasVariables(text: string): boolean {
    if (!text) {
        return false;
    }
    
    return VARIABLE_PATTERN.test(text);
}

/**
 * Validates variable names in text to ensure they follow naming conventions.
 * Variable names must start with a letter or underscore and contain only
 * letters, numbers, and underscores.
 * 
 * @param text - The text to validate
 * @returns Object with validation result and any invalid variable names found
 * 
 * @example
 * validateVariables("Hello #{Name} and #{123Invalid}")
 * // Returns: { valid: false, invalidVars: ["123Invalid"] }
 */
export function validateVariables(text: string): {
    valid: boolean;
    invalidVars: string[];
} {
    // This pattern matches ANY #{...} including invalid ones
    const anyVariablePattern = /#\{([^}]+)\}/g;
    const validNamePattern = /^[a-zA-Z_][a-zA-Z0-9_]*$/;
    
    const invalidVars: string[] = [];
    const matches = Array.from(text.matchAll(anyVariablePattern));
    
    for (const match of matches) {
        const varName = match[1];
        if (!validNamePattern.test(varName)) {
            invalidVars.push(varName);
        }
    }
    
    return {
        valid: invalidVars.length === 0,
        invalidVars
    };
}

/**
 * Preview text with variables resolved, useful for showing users
 * what the final text will look like with sample data.
 * 
 * @param text - The text template
 * @param sampleData - Sample variable values for preview
 * @returns Resolved text with indication of missing variables
 * 
 * @example
 * previewText("Welcome #{Name}, email: #{Email}", { Name: "Alice" })
 * // Returns: "Welcome Alice, email: [Email: Not set]"
 */
export function previewText(
    text: string,
    sampleData: Record<string, any> = {}
): string {
    return resolveVariables(text, sampleData, {
        defaultValue: '[Not set]',
        removeUnresolved: false
    });
}

/**
 * Batch resolve variables in multiple text strings.
 * Useful for resolving variables in a set of messages or templates at once.
 * 
 * @param texts - Array of text strings to resolve
 * @param userData - Variable values to use for resolution
 * @param options - Resolution options
 * @returns Array of resolved text strings
 * 
 * @example
 * batchResolveVariables(
 *   ["Hello #{Name}", "Email: #{Email}"],
 *   { Name: "Bob", Email: "bob@example.com" }
 * )
 * // Returns: ["Hello Bob", "Email: bob@example.com"]
 */
export function batchResolveVariables(
    texts: string[],
    userData: Record<string, any>,
    options: VariableResolverOptions = {}
): string[] {
    return texts.map(text => resolveVariables(text, userData, options));
}
